lec02-basic-r.Rmd

```diff
@@ -1,4 +1,765 @@
+These are the basic building blocks that all R objects are built from. The other
+4 **atomic vector** types are:
+
+* `"logical"` for `TRUE` and `FALSE` (the boolean data type)
+* `"integer"` for integer numbers (e.g., `2L`, the `L` indicates to R that it's an integer)
+* `"complex"` to represent complex numbers with real and imaginary parts (e.g.,
+  `1 + 4i`) and that's all we're going to say about them
+* `"raw"` for bitstreams that we won't discuss further
+
+Vectors are one of the many **data structures** that R uses. Other important
+ones are lists (`list`), matrices (`matrix`), data frames (`data.frame`),
+factors (`factor`) and arrays (`array`). In this class, we will focus on data
+frames, which is most commonly used one for data analyses.
+
+> **Challenge**
+>
+> * We’ve seen that atomic vectors can be of type character, numeric (or
+double), integer, and logical. But what happens if we try to mix these types in 
+>   a single vector? Find out by using `class` to test these examples.
+
+> ```r
+> num_char <- c(1, 2, 3, 'a')
+> num_logical <- c(1, 2, 3, TRUE)
+> char_logical <- c('a', 'b', 'c', TRUE)
+> tricky <- c(1, 2, 3, '4')
+> ```
+
+> <!-- * _Answer_: R implicitly converts them to all be the same type -->
+>
+
+> - This happens because vectors can be of only one data type. Instead of
+>  throwing an error and saying that you are trying to mix different types in
+>  the same vector, R tries to convert (coerce) the content of this vector to
+>  find a "common denominator". A logical can be turn into 1 or 0, and a number
+>  can be turned into a string/character representation. It would be difficult
+>  to do it the other way around: would 5 be TRUE or FALSE? What number would
+>  't' be?
+
+> - In R, we
+>   call converting objects from one class into another class
+>   _coercion_. These conversions happen according to a hierarchy,
+>   whereby some types get preferentially coerced into other
+>   types. Can you draw a diagram that represents the hierarchy of how
+>   these data types are coerced?
+> <!-- * _Answer_: `logical -> numeric -> character <-- logical` -->
+
+
+```{r, echo=FALSE, eval=FALSE}
+num_char <- c(1, 2, 3, "a")
+
+num_logical <- c(1, 2, 3, TRUE)
+
+char_logical <- c("a", "b", "c", TRUE)
+
+tricky <- c(1, 2, 3, "4")
+```
+
+
+## Subsetting vectors
+
+If we want to extract one or several values from a vector, we must provide one
+or several indices in square brackets. For instance:
+
+```{r}
+animals <- c("mouse", "rat", "dog", "cat")
+animals[2]
+animals[c(3, 2)]
+```
+
+We can also repeat the indices to create an object with more elements than the
+original one:
+
+```{r}
+more_animals <- animals[c(1, 2, 3, 2, 1, 4)]
+more_animals
+```
+
+R indices start at 1. Programming languages like Fortran, MATLAB, Julia, and R
+start counting at 1, because that's what human beings typically do. Languages in
+the C family (including C++, Java, Perl, and Python) count from 0 because that's
+simpler for computers to do.
```

Comment:
I don't think simplicity for computers is the reason.
https://python-history.blogspot.ca/2013/10/why-python-uses-0-based-indexing.html
https://www.cs.utexas.edu/~EWD/transcriptions/EWD08xx/EWD831.html

Reply:
Seems like it was when the initial implementation was done https://en.wikipedia.org/wiki/Zero-based_numbering#Origin. Will add "historically".

Reply:
Also added that it can allow for more elegant code.